### ISSUE TITLE
Connection pool? Use connection multiple times out of context creation

### ISSUE THREAD

**Issue by shsimeonova** [ronf/asyncssh#270]
I have the following code:

```
class SFTP:
    open_connections = {}

    @classmethod
        async def connect(cls, credentials):
            async with asyncssh.connect(**credentials) as conn:
                conn.set_keepalive()
                cls.open_connections[credentials['host']] = conn
                return conn
```

and i have the following methods:

```
@classmethod
    async def remote_destination_exists(cls, remote_path, credentials):
        async with asyncssh.connect(**credentials) as conn:
            sftpclient = await conn.start_sftp_client()
            exists = await sftpclient.exists(path=remote_path)

            return exists
```
and
```
@classmethod
    async def get_file(cls, remote_path, credentials):
        host_name = credentials['host']
        if host_name not in cls.open_connections or cls.open_connections[host_name] is None:
            connection = await cls.connect(credentials)
        else:
            connection = cls.open_connections[host_name]

        sftpclient = await connection.start_sftp_client()
        try:
            file = await sftpclient.open(path=remote_path, encoding=None)
            content = await file.read()
            return content
        except IOError:
            return False
```

and i want to replace the connection creation in remote_destination_exists with a connection thats being obtained from a pool and used in the get_file method later

I read one of your answers in https://github.com/ronf/asyncssh/issues/172 to create a dictionary that keeps connections open by their host as a key but i keep getting the asyncssh.misc.ChannelOpenError: SSH connection closed error. Im missing something but cant really understand what, can you please provide examples for this?

**Comment by ronf** [comment]
You're close here. The problem is mainly in your use of "async with" in the connect() function. That closes the connection as soon as it exits the block, so you can't return "conn" there and expect to be able to use it in the calling function. You need to just do a plain assignment on the call to asyncssh.connect().

I rewrote your connect() function to build in the logic to look for already-open connections, and added some "async with" calls to clean up the SFTP client and SFTP file objects. I also adjusted your check for IOError to look for asyncssh.SFTPError instead, and also adjusted the error path to return `None` instead of `False` (purely a personal preference for how to handle "optional" return values. The result looks as follows:

```python
class SFTP:
    open_connections = {}

    @classmethod
    async def connect(cls, credentials):
        host = credentials['host']
        conn = cls.open_connections.get(host)

        if not conn:
            conn = await asyncssh.connect(**credentials)
            cls.open_connections[host] = conn

        return conn

    @classmethod
    async def get_file(cls, remote_path, credentials):
        conn = await cls.connect(credentials)

        async with conn.start_sftp_client() as sftp_client:
            try:
                async with sftp_client.open(remote_path, encoding=None) as f:
                    return await f.read()
            except asyncssh.SFTPError:
                return None
```

If you wanted to, you could actually reuse the SFTP client objects rather than just the connection objects, if you're not trying to do multiple file transfers in parallel. The approach would be similar, but you'd move the start_sftp_client() call into the connect() function and have it save those objects in the dictionary instead of the connection objects. This would look something like:

```python
class SFTP:
    sftp_clients = {}

    @classmethod
    async def connect(cls, credentials):
        host = credentials['host']
        sftp_client = cls.sftp_clients.get(host)

        if not sftp_client:
            conn = await asyncssh.connect(**credentials)
            sftp_client = await conn.start_sftp_client()
            cls.sftp_clients[host] = sftp_client

        return sftp_client

    @classmethod
    async def get_file(cls, remote_path, credentials):
        sftp_client = await cls.connect(credentials)

        try:
            async with sftp_client.open(remote_path, encoding=None) as f:
                return await f.read()
        except asyncssh.SFTPError:
            return None
```

**Comment by shsimeonova** [comment]
Thanks you for your quick response and efforts!

**Comment by shsimeonova** [comment]
Hey, Ron

I can't fully understand from the docs, how are SSHClientConnections being closed? With this implementation if one request from the app closes the connection, other requests using this connection will fail if they've obtained the connection before it got closed. 

If i take the approach with SFTPClients and call exit() on them, how long is the connection going to stay opened? If another request uses the same sftp client and i exit the opened session will the session fail with the same success if i use connections?

**Comment by ronf** [comment]
If you are calling exit() yourself, all you'd need to do is remove the corresponding SFTPClient from the sftp_clients dictionary. However, if the connection is closed by the SFTP server, it's a bit trickier.

If you try to use an existing SFTPClient object previously opened on that connection, it will most likely return an error of either FX_NO_CONNECTION if it has already detected the connection close or FX_CONNECTION_LOST If the close happens while processing a new request. One option would be to retry the request in these cases. For instance:

```python
class SFTP:
    sftp_clients = {}

    @classmethod
    async def connect(cls, credentials):
        host = credentials['host']
        sftp_client = cls.sftp_clients.get(host)

        if not sftp_client:
            conn = await asyncssh.connect(**credentials)
            sftp_client = await conn.start_sftp_client()
            cls.sftp_clients[host] = sftp_client

        return sftp_client

    @classmethod
    async def get_file(cls, remote_path, credentials):
        attempts = 0 

        while attempts < 3:
            sftp_client = await cls.connect(credentials)

            try:
                async with sftp_client.open(remote_path, encoding=None) as f:
                    return await f.read()
            except (asyncssh.SFTPNoConnection, asyncssh.SFTPConnectionLost):
                cls.sftp_clients.pop(credentials['host'], None)
                attempts += 1
            except asyncssh.SFTPError:
                return None
```

The connect() method is unchanged from above, but there's now some new logic to look for the SFTPNoConnection and SFTPConnectionLost exceptions. These exceptions are only available in the latest "develop" commit from earlier today, but you could get a similar result inside the "except asyncssh.SFTPError" if you test exc.code for FX_NO_CONNECTION or FX_CONNECTION_LOST and then do a similar thing to remove the sftp_client from sftp_clients and retry.

**Comment by ronf** [comment]
Closing due to inactivity. Please open a new issue if you have additional questions.
